compiler/rustc_lint/src/unused.rs

```diff
@@ -293,6 +291,22 @@ impl<'tcx> LateLintPass<'tcx> for UnusedResults {
                     is_ty_must_use(cx, pinned_ty, expr, span)
                         .map(|inner| MustUsePath::Pinned(Box::new(inner)))
                 }
+                // Suppress warnings on `Result<(), UninhabitedType>` (e.g. `Result<(), !>`).
+                ty::Adt(def, args)
+                    if cx.tcx.is_diagnostic_item(sym::Result, def.did())
+                        && args.type_at(0).is_unit()
+                        && !args.type_at(1).is_inhabited_from(
+                            cx.tcx,
+                            parent_mod_did,
+                            cx.typing_env(),
+                        )
+                        && !matches!(
+                            args.type_at(1).kind(),
+                            ty::Adt(err_def, _) if err_def.is_variant_list_non_exhaustive()
```

Comment:
I believe `#[non_exhaustive]` is already handled by `is_inhabited_from`.

Reply:
That's because `#[non_exhaustive]` is ignored for local types from the current crate. It only applies when using the type from another crate.

I think we should be consistent with the behavior of other parts of the compiler that handle exhaustiveness checking in this regards.

Reply:
@Amanieu Yes, I did realize that; I had used `is_variant_list_non_exhaustive` intentionally in order to make it treat types from the *same* crate as inhabited if `non_exhaustive`. But you're right that if not trying to achieve that, I could leave it out and get the "`non_exhaustive` from a *different* crate" behavior for free.